spinta/cli/push.py

```diff
@@ -156,50 +168,153 @@ def push(
+                context.attach('push.state.conn', state.engine.begin)
+                _reset_pushed(context, models, state.metadata)
+
+            _push(
+                context,
+                client,
+                creds.server,
+                models,
+                rows,
+                state=state,
+                stop_time=stop_time,
+                stop_row=stop_row,
+                chunk_size=chunk_size,
+                dry_run=dry_run,
+                stop_on_error=stop_on_error,
+            )
+            repeat -= 1
+
+        _clean_up(
             context,
+            client,
+            creds.server,
             models,
-            counts,
-            limit,
+            state=state,
+            chunk_size=chunk_size,
+            dry_run=dry_run,
             stop_on_error=stop_on_error,
         )
-        rows = _prepare_rows_for_push(rows)
 
-        rows = tqdm.tqdm(rows, 'PUSH', ascii=True, total=sum(counts.values()))
 
-        if stop_time:
-            rows = _add_stop_time(rows, stop_time)
+class _State(NamedTuple):
+    engine: sa.engine.Engine
+    metadata: sa.MetaData
+
+
+def _push(
+    context: Context,
+    client: requests.Session,
+    server: str,  # https://example.com/
+    models: List[Model],
+    rows: Iterable[ModelRow],
+    *,
+    state: Optional[_State] = None,
+    stop_time: Optional[int] = None,    # seconds
+    stop_row: Optional[int] = None,     # stop aftern given number of rows
+    chunk_size: Optional[int] = None,   # split into chunks of given size in bytes
+    dry_run: bool = False,              # do not send or write anything
+    stop_on_error: bool = False,        # raise error immediately
+) -> None:
+    rows = _rows_to_push_rows(rows)
+
+    if stop_time:
+        rows = _add_stop_time(rows, stop_time)
+
+    if state:
+        rows = _check_push_state(context, rows, state.metadata)
 
-        if state:
-            rows = _check_push_state(context, rows, metadata)
+    rows = _prepare_rows_for_push(rows)
 
-        if stop_row:
-            rows = itertools.islice(rows, stop_row)
+    if stop_row:
+        rows = itertools.islice(rows, stop_row)
 
-        rows = _push_to_remote_spinta(rows, creds, chunk_size, dry_run=dry_run)
+    rows = _push_to_remote_spinta(
+        client,
+        server,
+        rows,
+        chunk_size,
+        dry_run=dry_run,
+    )
 
-        if state and not dry_run:
-            rows = _save_push_state(context, rows, metadata)
+    if state and not dry_run:
+        rows = _save_push_state(context, rows, state.metadata)
 
-        _push_rows(rows, stop_on_error)
+    _push_rows(rows, stop_on_error)
+
+
+def _clean_up(
+    context: Context,
+    client: requests.Session,
+    server: str,
+    models: List[Model],
+    *,
+    state: Optional[_State] = None,
+    chunk_size: Optional[int] = None,
+    dry_run: bool = False,
+    stop_on_error: bool = False,
+):
```

Comment:
Maybe it is possible to incorporate handling of deleted rows into may iterator? Because, ne this is a completely separate flow of code, and things like limiting number of rows is not added.

So instead of separate flow, I would suggest to move `_get_deleted_rows` into `iter_mode_rows` or probably create a separate `_read_rows`, where first we iterate over existing model rows and then deleted rows.

Then everything would fall into a single flow:

- `_read_rows` - first read available rows, then read deleted from state
- `_rows_to_push_rows` - wrap into `_Push row`
- `_add_stop_time` - limit whole push time
- `_check_push_state` - sync with push state
- `slice` - if limit is applied
- `_push_to_remote_spinta` - sync with remote
- `_save_push_state` - update push state

This way we will have a single iterator, and will be able to add more things on top of it, if needed.